Minerals.gd

```diff
@@ -0,0 +1,75 @@
+extends Spatial
+
+
+# Declare member variables here. Examples:
+enum {
+	NODE,
+	PASSED_TIME,
+	RESPAWN_TIME,
+	STATE
+}
+
+var scene = null
+var mineral = null
+var listBroken = [] # [[mineral, 0, 30, false]] in it 
+var thread
+
+# Called when the node enters the scene tree for the first time.
+func _ready():
+	scene = get_tree().get_root()
+	loadStateMinerals()
+	thread = Thread.new()
+	thread.start(self, "respawnMinerals", listBroken)
+
+# warning-ignore:shadowed_variable
+func respawnMinerals(listBroken):
+	while (true):
+		for i in range(0, listBroken.size()):
+			if (listBroken[i][STATE] == true
+				and OS.get_ticks_msec() - listBroken[i][PASSED_TIME] >= listBroken[i][RESPAWN_TIME]
+				and is_instance_valid(listBroken[i][NODE])):
+					listBroken[i][NODE].visible = true
+					listBroken[i][STATE] = false
+					listBroken.remove(i)
+					i -= 1
+
+func _unhandled_input(event):
+	if event is InputEventKey:
+		if event.pressed and event.scancode == KEY_ESCAPE:
+			get_tree().quit()
+		elif event.pressed and event.scancode == KEY_ENTER:
+			mineral = scene.get_node("Minerals/Mineral1")
+			changeStateMineral(mineral)
+
+func changeStateMineral(mineral : Node):
+	if mineral != null and mineral.visible == true:
+		listBroken.push_back([mineral, OS.get_ticks_msec(), 30000, true])
```

Comment:
Be careful when using an Array with a thread if both the main thead and the second thread access the same array and do some operations on it this could lead to race conditions, consider using a Mutex in those cases to protect your code from race conditions